File: firewall_manager/app/models/policy.py
from sqlalchemy import Column, Integer, String, ForeignKey, Boolean, DateTime
from sqlalchemy.orm import relationship
from app.db.session import Base
from datetime import datetime
from zoneinfo import ZoneInfo

class Policy(Base):
    __tablename__ = "policies"

    id = Column(Integer, primary_key=True, index=True)
    device_id = Column(Integer, ForeignKey("devices.id"))
    vsys = Column(String, nullable=True)
    seq = Column(Integer, nullable=True)
    rule_name = Column(String, index=True, nullable=False)
    enable = Column(Boolean(create_constraint=False), nullable=True)
    action = Column(String, nullable=False)
    source = Column(String, nullable=False)
    user = Column(String, nullable=True)
    destination = Column(String, nullable=False)
    service = Column(String, nullable=False)
    application = Column(String, nullable=True)
    security_profile = Column(String, nullable=True)
    category = Column(String, nullable=True)
    description = Column(String, nullable=True)
    # 정책 사용이력의 마지막 히트 시간
    last_hit_date = Column(DateTime, nullable=True)
    is_active = Column(Boolean(create_constraint=False), default=True, nullable=False)
    # 시스템 시간(한국시간)으로 저장하기 위해 default는 런타임에서 주입
    last_seen_at = Column(DateTime, default=lambda: datetime.now(ZoneInfo("Asia/Seoul")).replace(tzinfo=None), nullable=False)

    device = relationship("Device")

File: frontend/src/types/policy.ts
export interface Policy {
  id: number
  device_id: number
  rule_name: string
  source: string
  destination: string
  service: string
  action: string
  vsys: string | null
  seq: number | null
  enable: boolean
  user: string | null
  application: string | null
  security_profile: string | null
  category: string | null
  description: string | null
  last_hit_date: string | null
  hit_count: number | null
  is_active: boolean
  last_seen_at: string | null
}

export interface FilterLeafNode {
  type: 'LEAF'
  field: string
  operator: 'contains' | 'equals' | 'not_equals' | 'not_contains' | 'gte' | 'lte' | 'only_within'
  value: string
}
export interface FilterGroupNode {
  type: 'AND' | 'OR'
  children: FilterExprNode[]
}
export type FilterExprNode = FilterLeafNode | FilterGroupNode

export interface PolicySearchRequest {
  device_ids: number[]
  vsys?: string
  vsys_negate?: boolean
  rule_name?: string
  rule_name_negate?: boolean
  action?: string
  action_negate?: boolean
  enable?: boolean
  user?: string
  user_negate?: boolean
  application?: string
  application_negate?: boolean
  security_profile?: string
  category?: string
  description?: string
  description_negate?: boolean
  last_hit_date_from?: string
  last_hit_date_to?: string
  src_ip?: string
  dst_ip?: string
  protocol?: string
  port?: string
  src_ips?: string[]
  dst_ips?: string[]
  src_ips_exact?: string[]
  dst_ips_exact?: string[]
  services?: string[]
  src_names?: string[]
  dst_names?: string[]
  service_names?: string[]
  src_ips_exclude?: string[]
  dst_ips_exclude?: string[]
  src_ips_exact_exclude?: string[]
  dst_ips_exact_exclude?: string[]
  services_exclude?: string[]
  src_names_exclude?: string[]
  dst_names_exclude?: string[]
  service_names_exclude?: string[]
  skip?: number
  limit?: number
  filter_expression?: FilterExprNode
}

export interface ChangeLogEntry {
  id: number
  device_id: number
  object_name: string
  action: 'created' | 'updated' | 'deleted' | 'hit_date_updated'
  timestamp: string | null
}

export interface PolicyHistoryEntry extends ChangeLogEntry {
  details: Record<string, unknown> | string | null
}

export interface PolicySearchResponse {
  policies: Policy[]
  valid_object_names: string[]
}
